### ISSUE TITLE
Profiling shader compiles

### ISSUE THREAD

**Issue by tsili** [cbaggers/cepl#255]
Where could I plug in into cepl to see where all the compile time goes. For example to see how long it took for varjo to generate the glsl and how long for cepl to compile that and how long for the driver to do its thing. Currently this is the best I got and what strikes a little odd to me is the growth rate of the compile times.

```
node count     Compile time     evaluation time for first pulse
100                 3                       6
200                 20                      30
300                 65                      75
400                 151                     157
500                 290                     282
600                 490                     475
700                 762                     735
800                 1113                    1053
```

I would like to know where to hack in some timers to start finding out why are the compile times almost doubling for each 100 new nodes.

```
Testing with model size 100

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  3.041 seconds of real time
  3.967538 seconds of total run time (3.966541 user, 0.000997 system)
  130.48% CPU
  149 forms interpreted
  163 lambdas converted
  10,319,106,454 processor cycles
  495,139,840 bytes consed

"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  5.780 seconds of real time
  6.662023 seconds of total run time (6.652977 user, 0.009046 system)
  115.26% CPU
  19,615,170,554 processor cycles
  672,522,800 bytes consed


Testing with model size 200

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  19.982 seconds of real time
  20.870541 seconds of total run time (20.853585 user, 0.016956 system)
  [ Run times consist of 0.083 seconds GC time, and 20.788 seconds non-GC time. ]
  104.45% CPU
  149 forms interpreted
  163 lambdas converted
  67,811,171,786 processor cycles
  3,106,125,504 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  27.586 seconds of real time
  29.539649 seconds of total run time (29.507746 user, 0.031903 system)
  [ Run times consist of 0.068 seconds GC time, and 29.472 seconds non-GC time. ]
  107.08% CPU
  93,614,211,678 processor cycles
  3,085,017,856 bytes consed

Testing with model size 300

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  65.239 seconds of real time
  65.256901 seconds of total run time (64.582948 user, 0.673953 system)
  [ Run times consist of 0.426 seconds GC time, and 64.831 seconds non-GC time. ]
  100.03% CPU
  149 forms interpreted
  163 lambdas converted
  221,391,446,318 processor cycles
  9,769,012,528 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  73.369 seconds of real time
  75.373209 seconds of total run time (75.302119 user, 0.071090 system)
  [ Run times consist of 0.109 seconds GC time, and 75.265 seconds non-GC time. ]
  102.73% CPU
  248,986,798,386 processor cycles
  9,244,122,480 bytes consed

Testing with model size 400

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  151.094 seconds of real time
  151.137313 seconds of total run time (151.095348 user, 0.041965 system)
  [ Run times consist of 0.175 seconds GC time, and 150.963 seconds non-GC time. ]
  100.03% CPU
  149 forms interpreted
  163 lambdas converted
  512,752,183,516 processor cycles
  22,408,958,112 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  154.922 seconds of real time
  157.037775 seconds of total run time (156.951759 user, 0.086016 system)
  [ Run times consist of 0.139 seconds GC time, and 156.899 seconds non-GC time. ]
  101.37% CPU
  525,742,357,612 processor cycles
  20,930,222,064 bytes consed

Testing with model size 500

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  289.905 seconds of real time
  289.987524 seconds of total run time (289.915504 user, 0.072020 system)
  [ Run times consist of 0.271 seconds GC time, and 289.717 seconds non-GC time. ]
  100.03% CPU
  149 forms interpreted
  163 lambdas converted
  983,818,189,706 processor cycles
  42,696,772,240 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  282.366 seconds of real time
  282.845360 seconds of total run time (282.080342 user, 0.765018 system)
  [ Run times consist of 0.664 seconds GC time, and 282.182 seconds non-GC time. ]
  100.17% CPU
  958,234,426,184 processor cycles
  39,671,882,128 bytes consed

Testing with model size 600

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  489.712 seconds of real time
  491.392494 seconds of total run time (490.554430 user, 0.838064 system)
  [ Run times consist of 0.620 seconds GC time, and 490.773 seconds non-GC time. ]
  100.34% CPU
  149 forms interpreted
  163 lambdas converted
  1,661,882,144,808 processor cycles
  73,333,848,736 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  473.982 seconds of real time
  476.254795 seconds of total run time (475.292807 user, 0.961988 system)
  [ Run times consist of 0.646 seconds GC time, and 475.609 seconds non-GC time. ]
  100.48% CPU
  1,608,502,247,318 processor cycles
  68,024,845,712 bytes consed

Testing with model size 700

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  761.911 seconds of real time
  762.146622 seconds of total run time (760.288702 user, 1.857920 system)
  [ Run times consist of 1.435 seconds GC time, and 760.712 seconds non-GC time. ]
  100.03% CPU
  149 forms interpreted
  163 lambdas converted
  2,585,614,719,840 processor cycles
  114,807,390,688 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  734.478 seconds of real time
  736.841229 seconds of total run time (735.691283 user, 1.149946 system)
  [ Run times consist of 0.694 seconds GC time, and 736.148 seconds non-GC time. ]
  100.32% CPU
  2,492,517,967,558 processor cycles
  106,334,035,200 bytes consed

Testing with model size 800

"Creating new model."
; recompile cpu side of (MODEL-PIPELINE ...)
Evaluation took:
  1113.312 seconds of real time
  1113.678566 seconds of total run time (1110.259724 user, 3.418842 system)
  [ Run times consist of 1.508 seconds GC time, and 1112.171 seconds non-GC time. ]
  100.03% CPU
  149 forms interpreted
  163 lambdas converted
  3,778,127,183,534 processor cycles
  171,264,476,448 bytes consed


"Evaluating model."
; uploading (MODEL-PIPELINE ...)

NIL
Evaluation took:
  1051.559 seconds of real time
  1054.126414 seconds of total run time (1051.151391 user, 2.975023 system)
  [ Run times consist of 1.699 seconds GC time, and 1052.428 seconds non-GC time. ]
  100.24% CPU
  3,568,560,603,608 processor cycles
  158,598,094,352 bytes consed
```

**Comment by cbaggers** [comment]
Im very willing to believe that varjo is too slow, I've never done any work on making that quick (though funnily I was just thinking about this before I saw this ticket). Also CEPL is trying to propagate changes from gpu-functions (which means recompiling all gpu-funcs & pipelines that use it) That is going to have some nasty speed repercussions as more functions are in the system.

What are nodes & pulses in this case? I'd be quite interested in your test setup.

`cepl.pipelines` Is where all the logic for compilation is. `rolling-translate` is the function from varjo that cepl  calls to compile things.

Thanks for looking into this. Im a tad nervous as I know how messy a lot of this stuff is but also excited with the idea that we can make it faster

**Comment by cbaggers** [comment]
ok so I made a quick test. https://github.com/cbaggers/varjo/blob/feature/perf/protocode/profile.lisp

I run `profile-all-varjo` which instruments all the functions. I then call `gooo` passing in a list of stages to be compiled together. it runs this 30 times so we get enough data. I then dumped out the results here https://github.com/cbaggers/varjo/blob/feature/perf/protocode/varjo-perf-issues.ods

The second page has the graphs of some of the worst offenders. But the results are not too surprising. I can see things that will be easy for me to start looking at and it might be that those things alone will be enough to get some better build times.

**Comment by cbaggers** [comment]
This was my hacky test
```
(gooo
   (mapcar #'varjo.internals:starting-stage 
			   (v-compile
				'((tess-level-inner :float) (tess-level-outer :float)
				  (projection :mat4) (model->clip :mat4) (normal-mat :mat3)
				  (light-position :vec3) (diffuse-material :vec3) (ambient-material :vec3))
				:410
				:draw-mode '(:patch 3)

				:vertex
				'(((position :vec4))
				  (values position
				   (s~ position :xyz)))

				:tessellation-control
				'(((position (:vec3 3)))
				  (declare (output-patch :vertices 3))
				  (let ((tess-level-inner 5f0)
						(tess-level-outer 5f0))
					(when (= gl-invocation-id 0)
					  (setf (aref gl-tess-level-inner 0) tess-level-inner
							(aref gl-tess-level-outer 0) tess-level-outer
							(aref gl-tess-level-outer 1) tess-level-outer
							(aref gl-tess-level-outer 2) tess-level-outer)))
				  (aref position gl-invocation-id))

				:tessellation-evaluation
				'(((position (:vec3 3)))
				  (declare (tessellate-to
							:primitive :triangles
							:spacing :equal
							:order :ccw))
				  (let* ((p0 (* (x gl-tess-coord) (aref position 0)))
						 (p1 (* (y gl-tess-coord) (aref position 1)))
						 (p2 (* (z gl-tess-coord) (aref position 2)))
						 (pos (normalize (+ p0 p1 p2))))
					(values
					 (* model->clip (v! pos 1))
					 pos
					 gl-tess-coord)))

				:geometry
				'(((position (:vec3 3)) (patch-distance (:vec3 3)))
				  (declare (output-primitive :kind :triangle-strip
							:max-vertices 3))
				  (let* ((a (- (aref position 2) (aref position 0)))
						 (b (- (aref position 2) (aref position 0)))
						 (facet-normal (* normal-mat (normalize (cross a b)))))
					(emit ()
						  (gl-position (aref gl-in 0))
						  (aref patch-distance 0)
						  facet-normal
						  (v! 1 0 0))
					(emit ()
						  (gl-position (aref gl-in 1))
						  (aref patch-distance 1)
						  facet-normal
						  (v! 0 1 0))
					(emit ()
						  (gl-position (aref gl-in 2))
						  (aref patch-distance 2)
						  facet-normal
						  (v! 0 0 1))
					(end-primitive)
					(values)))

				:fragment
				'(((patch-distance :vec3) (facet-normal :vec3) (tri-distance :vec3))
				  (labels ((amplify ((d :float) (scale :float) (offset :float))
							 (let* ((d (+ (* scale d) offset))
									(d (clamp d 0 1)))
							   (- 1 (expt (* d d -2) 2)))))
					(let* ((light-position (v! 0 1 0))
						   (diffuse-material (v! 1 0 0))
						   (ambient-material (v! 0.2 0.2 0.2))
						   ;;
						   (n (normalize facet-normal))
						   (l light-position)
						   (df (abs (dot n l)))
						   (color (+ ambient-material (* df diffuse-material)))
						   (d1 (min (min (x tri-distance) (y tri-distance))
									(z tri-distance)))
						   (d2 (min (min (x patch-distance) (y patch-distance))
									(z patch-distance)))
						   (color (* (+ ambient-material (* df diffuse-material))
									 (amplify d1 40 -0.5)
									 (amplify d2 50 -0.5))))
					  (v! color 1))))))
```

**Comment by tsili** [comment]
Haven't taken a look at the results you got yet, but will when I get the time. In the mean time, let me answer some of your questions.

The nodes are the base units of the GP system, which means in this case procedures from which the models are constructed from. A simple example, however, useless one of a node would be in glsl as

```
void LIB_EXPR_0(MODEL DATA)
{
    if ((DATA.NODES[int(gl_WorkGroupID.x)].INPUTS[0] == 58))
    {
        for (int I = 0;(I < 10);(++ I))
        {
            int OUTPUT_SLOT = DATA.NODES[int(gl_WorkGroupID.x)].OUTPUTS[I];
            int OUTPUT_NODE = int(trunc(float((OUTPUT_SLOT / 10))));
            DATA.NODES[OUTPUT_NODE].INPUTS[int(mod(float(OUTPUT_SLOT),float(10)))] = 42;
        }
    }
    return;
}
```
This is something the system did write on its own and fed to varjo.

In the models I am constructing, I have currently three ways to represent these nodes; As individual functions as depicted above, as inline code withing a call-tree or as a labels section preceding the call tree. I have an example call-tree depicted below. It is very inefficient and I should turn that to a binary tree instead, but it works.

```
void MODEL_SHADER()
{
    int NODE_ID = int(gl_WorkGroupID.x);
    if ((NODE_ID == 0))
    {
        LIB_EXPR_0(DATA);
    }
    else
    {
        if ((NODE_ID == 1))
        {
            LIB_EXPR_1(DATA);
        }
        else
        {
            if ((NODE_ID == 2))
            {
                LIB_EXPR_2(DATA);
            }
            else
            {
                if ((NODE_ID == 3))
                {
                    LIB_EXPR_3(DATA);
                }
                else
                {
                    if ((NODE_ID == 4))
                    {
                        LIB_EXPR_4(DATA);
                    }
                    else
                    {
                        if ((NODE_ID == 5))
                        {
                            LIB_EXPR_5(DATA);
                        }
                        else
                        {
                            if ((NODE_ID == 6))
                            {
                                LIB_EXPR_6(DATA);
                            }
                            else
                            {
                                if ((NODE_ID == 7))
                                {
                                    LIB_EXPR_7(DATA);
                                }
                                else
                                {
                                    if ((NODE_ID == 8))
                                    {
                                        LIB_EXPR_8(DATA);
                                    }
                                    else
                                    {
                                        if ((NODE_ID == 9))
                                        {
                                            LIB_EXPR_9(DATA);
                                        }
                                        else
                                        {
                                            false;
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return;
}
```

Pulses in this case are the events of calling map-g on the model shader with some input data. In the test depicted in the first post, the data was a simple blank buffer. In this case all nodes are gone through and nothing happens. Therefore, the majority of the time taken by the first pulse, is the time it takes to upload the model to the gpu.

Having the three ways to compile the model, I decided to do a little test to find what way would be the most efficient. I set for my computer to compile 12 models; four with all three ways. The models compiled were of sizes 500, 1000, 1500 and 2000. Originally I thought that the method of separating the node functions outside of the main shader would be the fastest as there would be four threads compiling the nodes. And I thought that the inline definition would be the second fastest as less code would be written than with the labels section. However, here are the results.

```
nodes   inline   labels   modular
 500       .6      .15      .4
1000      5       1.2      3.6
1500     13.8     3.8     10.4
2000      DNF     8.1      DNF
```
The compile times are in hours. As you can see, I was incorrect in my assumptions. The questions this unexpected order raises come to how cepl and varjo handle the shaders. Firstly you stated that cepl may be recompiling every shader every time they are needed, or so I understood, and this explains the result gotten for the modular approach. However, this does not explain why having a separate labels section lisp side as opposed to defining every node inline would be that much faster. The other hypothesis I get leads me to ask, how does varjo handle whitespace. Is the shader compiled into a compact format and spaced out and indented when called for printing, or is it compiled into this human friendly format? This question comes from the fact that in the inline function definition model, the node functions quite literally have thousands of levels of indentation, which looks quite trippy in emacs when looking at it.

When running the models, I have ten threads to use for compiling, one to manage the model on the gpu, and the last one to pre-process input data. If I want to have each model be tested for an hour, I need to have the compilation of the next one take at maximum ten hours. Or in other words have one model ready every hour. With model size of 2000 and using the labels section, this should not pose a problem. However, my initial calculations show that the models would end up in the size class of 10000+ nodes and there I get the problem.

I'll use those profiling functions you showed sometime soon and see what kind of results I get. And I would like to know if I can force cepl to not compile any shaders unless separately asked to and see if something changes.

**Comment by cbaggers** [comment]
Thanks for all the info. I think I understand more of this now

A node is a function. The function can be inlined, a local function (using labels) or 'modular' which (I think) is a separate CEPL gpu-function (defined with `defun-g`). So when you say 2000 nodes (in the timing information) it's 2000 functions.

Also I wanted to ask: When you `labels section` are you taking about [labels](http://www.lispworks.com/documentation/HyperSpec/Body/s_flet_.htm) or something else? I'm not used to the terminology of [section](http://www.lispworks.com/documentation/HyperSpec/Body/26_glo_s.htm#section) in lisp being used in this context.

> However, this does not explain why having a separate labels section lisp side as opposed to defining every node inline would be that much

To answer the inlining question I'd need to know how you are doing the inlining. CEPL & Varjo don't have `inline` declarations so I guess you are using gpu-macros (or generating the inline code yourself?). Any info here will help me find out why that is slower.

>  Firstly you stated that cepl may be recompiling every shader every time they are needed,

Not quite, sorry if I was confusing. If a gpu-function is compiled it will trigger the recompilation of functions/pipelines that use it. This is so we get live update when modifying our GPU code as it is running. (there is also going to be a final compilation of the whole pipeline which may be deferred to when the pipeline is first called, this is loosely defined so I have the option to do it at compile time if I want)

In summary:

When a `defun-g` is compiled it:
- calls varjo to test compile the code. This makes sure there is at least 1 shader stage this code would be valid in.
- recompiles any gpu-functions that use this function
- recompiles any pipelines that use this function as a stage

Also. When a `defpipeline` is recompiled:
- regenerates it's lisp code, but doesn call varjo

When a pipeline is first called
- Varjo is called to compile the stages and the results are cached.



> leads me to ask, how does Varjo handle whitespace

Regarding style of generated glsl, like you have seen we emit human readable glsl. There is very little advantage to us minifying it and also thousands of levels of indentations sound like really slow GPU code to be honest. GPUs, in general, don't like lots of branching so those huge levels of nesting sounds concerning (but this is runtime performance and shouldnt make a huge difference in compilation time). As you are branching based on an `int` maybe a switch statement would be better:

```
TESTS> (glsl-code
		(compile-frag ((a :int :flat)) :450 nil
		  (let (((b :vec4) ))
			(switch a
			  (0 (setf b (vec4 1)))
			  (1 (setf b (vec4 2)))
			  (2 (setf b (vec4 3)))
			  (3 (setf b (vec4 4)))
			  (4 (setf b (vec4 5)))
			  (5 (setf b (vec4 6)))
			  (6 (setf b (vec4 7)))
			  (7 (setf b (vec4 8)))
			  (8 (setf b (vec4 9)))
			  (9 (setf b (vec4 10))))
			b)))
"// fragment-stage
#version 450

in _IN_BLOCK_
{
     flat in int A;
} v_in;

layout(location = 0)  out vec4 _FRAGMENT_STAGE_OUT_0;

void main()
{
    vec4 B;
    
    switch (v_in.A) {
        case 0:
        B = vec4(float(1));
        break;
    
        case 1:
        B = vec4(float(2));
        break;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             MORE CODE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        case 9:
        B = vec4(float(10));
        break;
    }
    vec4 g_GEXPR0_15834 = B;
    _FRAGMENT_STAGE_OUT_0 = g_GEXPR0_15834;
    return;
}"
```
`switch` returns `void` in Vari as it's a statement in glsl, that is why we need to assign the result to a local variable.

> as there would be four threads compiling the nodes
I'm very confused by this. CEPL does not create any threads, and does not multithread compilation.

**Comment by tsili** [comment]
Good that we can get to more even ground on this and I'm sorry for my poor terminology, I'm not majoring in computer science, but in finance.

So, indeed a node is a function that can be defined in three ways in the code. The 'modular' being a separate gpu function defined with `defun-g`, inline being expanded using defmacro-g for all of the functionality that the nodes are allowed to have and with [labels](http://www.lispworks.com/documentation/HyperSpec/Body/s_flet_.htm), having me refer to different parts of the main function as sections with no further meaning. The term 'section' simply stems from the number of research papers I have had to read.

About using the switch statement, I should probably do that, but haven't yet gotten that far in my development of the system to have time to worry about the runtime performance. The whole system doesn't run at all yet. I was just using a really large cond tree as it allowed me to start working on the next part, which still is getting the data to the model and get something to go through all the way to the end. After a day I can then leave the computer to do some test overnight about applying what I find into the large structure I intend to run this system with.

Last night I tried to improve on the linear cond tree with this binary tree
```
(defmacro ncond (id-var &rest clauses)
  (if (rest clauses)
    (let* ((half-point (truncate (/ (length clauses)
                                    2.0)
                                 1))
           (last-half (subseq clauses half-point)))
      `(if (< ,(caar last-half) ,id-var)
           (ncond ,id-var ,@(subseq clauses 0 half-point))
         (ncond ,id-var ,@last-half)))
    (cadar clauses)))
```
I ran the test with the currently best found method of using labels, but I should probably try with the switches too. Any way here is what I got.

```
nodes       compile-time
 500          15   s
1000           1.5 m
2000          12   m
3000          41   m
4000           1.7 h
5000           2.9 h
6000           4.8 h
```
As I did expect, the compile times are lower, but I did not expect for them to be this low. The only change being the switch to the ncond as opposed to regular cond. I'll have to see what happens with the switches. I did have some larger ones set up as well, but I ran into some limitations in sbcl. I'll need to do a custom ebuild for that, which shouldn't be too hard on gentoo. And yes I have to agree with you on that the whitespace shouldn't affect much with any type of normal use.

To get to the last question you posted about cepl not being multi-threaded, I know that and just made the treads on my side using `(bt:make-thread (lambda () (define-model-here)))` As those models were compiled for pure testing of compiling I wasn't that concerned about potential thread safety issues that could happen within cepl. If some problems do come up, we can always fix them.

The specific issue about four threads being used to compile the nodes comes from some machinery I set up with the modular models first compiling all of their nodes before compiling themselves. As I had setup the compiles of the modular models, four of them, to happen simultaneously and use partially the same nodes, resulted in all four threads compiling the functions needed by the first of the models. Then as the smallest model starts to compile itself, the rest three treads continue to compile their needed nodes. And the process continues as long as there are more nodes needed.

To get forward I must thank you for the description on what happens from defun-g onwards. The fact of varjo being called to test compile when the functions are defined with `defun-g` and then again when they are first needed with `(map-g 'pipeline)` does explain why the time for the initial pulse from the first post is so similar to the compile time as all that is done there is call map-g on the pipeline with empty data. If I understand correctly, this act of testing results in all of the functions being compiled twice, first to test and second to use. If that is indeed the case, would it be possible to tell cepl and varjo to not do a test compile? As with the system I'm writing, all of the shaders are written by a computer and invalid ones can just be discarded. It won't matter if the invalidness is captured during test compile or real. In a repl environment and development that testing can be quite useful, but when there are numerous large shaders, from which some are expected to fail, and throughput being the major issue, the double compiling doesn't sound beneficial.

Here is what I need to do for every model.
1. Generate the genome for the model.
2. Extract each gene from the genome and turn those into descriptions of nodes.
3. Pile the first half of each of those nodes together into a shader stage using a defun-g.
4. Link that shader stage to a pipeline using defpipeline-g. this needs to be done once only as the stage can be redefined.
5. Pile the second half of each  those nodes into the initial connections of the model using make-gpu-array
6. Call the pipeline on some input data using map-g.
7. Read the output using pull-g.

**Comment by cbaggers** [comment]
Notes to self:
Looking at [this]() two functions stand out:

- 5.2    VARJO.INTERNALS::TRY-TYPE-SPEC->TYPE
- 33.1   VARJO.INTERNALS::BASIC-ARG-MATCHP

sprof says that 33% of the time we are in basic-arg-matchp. The important point with these two functions is that most of the functions that sprof saw in the call stack were functions compiling some chunk of code. These two however, do not, and so they are close to the culprits of part of the issue. If we could make these faster it is likely we could make a significant impact on compile time as a whole.

Also both of these make sense as culprits. I'm going to have a think about how we could encode the type info to allow basic-arg-matchp to be faster.

**Comment by cbaggers** [comment]
### More Notes

Well, I was reading the profile incorrectly. The important bit was:
```
Evaluation took:
  12.343 seconds of real time
~~~~ but at the bottom ~~~~

     0.785 |      0.015 | 109,752,032 | 4,915,380 |            | Total

estimated total profiling overhead: 10.53 seconds

```

0.785 for compiling a 5 stage pipeline 30 times isnt amazing..but it's a lot better that what we are seeing in the issue, so I think I need to look at the CEPL side instead.

For another test I turned off profiling and compiled the pipeline (5 stages) 3000 times. It took 40 seconds. 

Again these are only indicators, not case closed, but it strongly suggests I need to look at CEPL for the answers

**Comment by cbaggers** [comment]
Ok so our string indenting code is dirt slow. I'm going to rewrite all of this

**Comment by cbaggers** [comment]
@tsili I havent pushed the fix yet, but for highly nested code the new version of the glsl generation will be at least 10 times faster. I'll let you know when it is in.

**Comment by cbaggers** [comment]
Yay! Done a bunch of fixes on the branch. I will merge tomorrow (I just need to run the test suite on a couple of different machines)

Here is an example

```
(defun yup ()
  ;;             generate a cond with 200 arms
  (let* ((code `(let ((a 1))
                  (flet ((boop ((x :int))
                           (float x)))
                    (cond
                      ,@(loop :for i :below 200 :collect
                           `((= a ,i) (boop ,i))))
                    (v! 1 2 3 4))))
         (stage (make-stage :fragment nil nil '(:450) (list code))))
    (time (translate stage))))
```      

### On master

> Evaluation took:
>   5.630 seconds of real time
>   5.632000 seconds of total run time (5.308000 user, 0.324000 system)

### On feature/perf

> Evaluation took:
>   0.167 seconds of real time
>   0.164000 seconds of total run time (0.164000 user, 0.000000 system)

**Comment by tsili** [comment]
Great to hear you got this much extra performance. I will do some tests on that too. Before then I could probably give you the results I got with the switch alternative with the profiling on. 
[results](https://drive.google.com/open?id=1Qm5t40wOvhoH7ytTBGpjpnIEIEvxm2C1)
On the first sheet are results from running several compiles in parallel for 10 minutes using my little hackish method. And on the other sheet are the results from running many compiles one after the other.

**Comment by cbaggers** [comment]
No problem with not majoring in CS, neither did I :) It's very cool to hear that CEPL being used for non graphical stuff too.

I do agree that it makes sense to test first before worrying about performance, however I'm wondering if the glsl will compile at all in the end or if the driver will give up. It will be interesting to hear.

When you mentioned `(bt:make-thread (lambda () (define-model-here)))` I'm wondering what is in `(define-model-here)`. Just a little warning, `defun-g` & `defpipeline-g` have to be 'top level forms' so should not be inside a lambda. If you are using `lambda-g` & `pipeline-g` then awesome, you are the first person other than me to be using them. However they would have to be used on the thread they are compiled on and so those threads are going to need their own contexts ([cepl.examples](https://github.com/cbaggers/cepl.examples/blob/master/examples/shared-context.lisp) has an example of this, sadly that part of CEPL is not documented yet)

Currently there is no way to avoid the test compile when making gpu-functions, I do agree this is a problem though so I have [opened a ticket for this](https://github.com/cbaggers/cepl/issues/256).

Thanks for the data, I will have a look and see what could be causing such high times for %get-free-glsl-name, likely-recursion-p & replace-substrings-in-name.

**Comment by cbaggers** [comment]
I've merged some fixes into master so hopefully you will see the speedup when you next pull.

**Comment by cbaggers** [comment]
@tsili Tonight I will be adding CL's `case` to varjo. This will compile to a `glsl` switch when it can. It also compiles faster than the nested `if` and `cond` do (even after the fixes) so this should help with your project.

**Comment by tsili** [comment]
Thank you so much. I had been looking for the `case` before I ended up with the huge `cond` tree.

I'll get some tests in once I have some extra compute time to use. I'm currently running tests on the data pipeline I have made. And to answer your question about the `(define-model-here)`, it definitely uses simple `defun-g` at this time within another function and not on the top-level. I'll need to look into that as I get into runtime testing.

Thank you for the pointers about using `lamda-g`, I should get by with the examples. I haven't gotten as far with lisp to know why some things absolutely need to be on the top-level, but I know I cannot get my model and node definitions there. Therefore, the `lambda-g` and `pipeline-g` will come in useful.

**Comment by tsili** [comment]
OK, get some more tests in with the new versions of cepl and varjo. I have to say these results are once again something unexpected.

```
nodes       time      average time per node
  100       0.312 s     0.003
 1000     100     s     0.100
 2000       9.7   m     0.291
 3000      34.7   m     0.694
 4000       1.35  h     1.215
 5000       2.48  h     1.786
 6000       4.35  h     2.610
 7000       6.31  h     3.245
 8000       8.89  h     4.001
 9000      12.4   h     4.960
10000      17.4   h     6.264
```
As you can see, the time per node is still increasing as the models get bigger. Furthermore, the time spent compiling didn't change much. However, there were some improvements in the RAM consumption dropping from occasionally hitting swap on a 16 GB system to using 7-8 GB. I'm not sure of what could cause such a memory consumption drop and not affect the compile time. I'll run some more tests with the profiling on.

Additionally, I did some improvements on how the nodes are represented using GP-feature extraction to compress the nodes' code. Here are the results after adding this system into the mix.


```
nodes     time      average time per node
 1000     39.50 s     0.04
 2000      4.35 m     0.13
 3000     12.02 m     0.24
 4000     29.00 m     0.44
 5000     55.78 m     0.67
 6000      1.56 h     0.93
 7000      2.44 h     1.26
 8000      3.56 h     1.60
 9000      4.85 h     1.94
10000      6.46 h     2.33
```
With this I can get that magical 10 000 nodes' compile time under 10 hours and, therefore, making it usable.

**Comment by tsili** [comment]
I think I found something that could be part of the problem.

The glsl seems to be generated by varjo with `format`, so I did a little test as to what happens when many small blocks are put together with it.

```
(time (let ((text ""))
           (loop repeat 1000
                do (setq text (format nil "~A~D" text (random 100))))
           t))
```
```
characters     time
  1000          0.003
 10000          0.236
100000         17.766
200000        290.169
```
then it starts to get out of hand really quickly. I also tested `with-output-to-string`
```
(time (progn 
        (with-output-to-string (stream)
          (loop repeat 1000 do (write (random 100) :stream stream)))
        t))
```
```
characters     time
    1000      0.000
   10000      0.002
  100000      0.014
 1000000      0.143
10000000      1.429
```
This seems to scale linearly.

Because my shaders are simply a really big list of switches, this explains why it takes so long to compile. Do you have some ideas about this?

**Comment by cbaggers** [comment]
If you could grant me access to the repo, or email me a zip of the whole project, I'd be happy to have a look and see where the time is going.

**Comment by tsili** [comment]
Sent you a google drive link.

**Comment by cbaggers** [comment]
@tsili Brilliant. And I have a solution. Your shader can be about 8 lines of lisp and doesnt need to be generated each time. I've been writing a little document as I did this so I could understand so I'll finish it and send it to you

**Comment by cbaggers** [comment]
But first I want to profile this big shader and see where Varjo is sucking...because man it is sucking :D

**Comment by cbaggers** [comment]
Hi again, @tsili I have the 1000 node version down to 0.45seconds & the 10000 node one down to 45seconds. There is still more to do but I'd like you to be able to try it. Please check out the `feature/refactor-bindings` branch and see if that helps you. I will run the test suite tomorrow and if this all looks good I'll get it merged in to master. Then I'l have to see what the next candidate for tuning is.

Thanks for your patience

**Comment by cbaggers** [comment]
Oh and remember to try flet instead of labels :)

**Comment by cbaggers** [comment]
I've merged the changes into master final time for 1000 node version was 0.3 seconds & for the 10000 21.7 seconds.

Of course these specific numbers are only on my machine, but the scale of the improvement relative to the original is what matters in this case

**Comment by tsili** [comment]
I'm also running some tests right now. The improvements are substantial. I just finished a test on 100 000 nodes in just 1.2 hours including not only the compilation of the model shader, but the construction of the node connections as well.

That is enough for me to only have one worker continuously churn out models. This would leave me with one thread for reading in data, one for compiling the models, one for managing the running model and a whopping nine threads to analyze the results these models give. With this setup I should be able to reduce the amount of GP-generations needed for the whole thesis as I will be able to compress multiple models into a single run and share the data and computing cycles between them and  make much better predictions on how to change the models for better performance.

Thank you so much.

**Comment by cbaggers** [comment]
My pleasure, thankyou for providing such good data to work with.
Whilst there are still performance issues to work on in general, I'm going to close this particular issue for now.
Feel free to open new issues if/when you find more perf problems (I expect similar issue for thousands of nested variable declarations) peace
